golem-api-grpc/proto/golem/apidefinition/api_definition.proto

```diff
@@ -60,30 +68,66 @@ enum HttpMethod {
   HEAD = 8;
 }
 
-message WorkerBinding {
-  golem.component.VersionedComponentId component = 1;
+// Used in api definition repo and needs to be backward compatible
+message GatewayBinding {
+  // Optional field to keep backward compatibility
+  optional golem.component.VersionedComponentId component = 1;
   optional golem.rib.Expr worker_name = 2;
-  golem.rib.Expr response = 3;
+  // Optional field to keep backward compatibility
+  optional golem.rib.Expr response = 3;
   optional golem.rib.Expr idempotency_key = 4;
-  optional WorkerBindingType type = 5;
+  // type discriminator to keep backward compatibility
+  optional GatewayBindingType binding_type = 5;
+  optional StaticBinding static_binding = 6;
+  optional Middleware middleware = 7;
 }
 
-message CompiledWorkerBinding {
-  golem.component.VersionedComponentId component = 1;
-  optional golem.rib.Expr worker_name = 2;
-  optional golem.rib.RibByteCode compiled_worker_name_expr = 3;
-  optional golem.rib.RibInputType worker_name_rib_input = 4;
-  golem.rib.Expr response = 5;
-  golem.rib.RibByteCode compiled_response_expr = 6;
-  golem.rib.RibInputType response_rib_input = 7;
-  optional golem.rib.Expr idempotency_key = 8;
-  optional golem.rib.RibByteCode compiled_idempotency_key_expr = 9;
-  optional golem.rib.RibInputType idempotency_key_rib_input = 10;
-  optional golem.rib.WorkerFunctionsInRib worker_functions_in_response = 11;
-  optional WorkerBindingType type = 12;
+// Used in api definition repo and needs to be backward compatible
+message CompiledGatewayBinding {
+    // Optional field to keep backward compatibility
+    optional golem.component.VersionedComponentId component = 1;
+    optional golem.rib.Expr worker_name = 2;
+    optional golem.rib.RibByteCode compiled_worker_name_expr = 3;
+    optional golem.rib.RibInputType worker_name_rib_input = 4;
+    optional golem.rib.Expr response = 5;
+    optional golem.rib.RibByteCode compiled_response_expr = 6;
+    optional golem.rib.RibInputType response_rib_input = 7;
+    optional golem.rib.Expr idempotency_key = 8;
+    optional golem.rib.RibByteCode compiled_idempotency_key_expr = 9;
+    optional golem.rib.RibInputType idempotency_key_rib_input = 10;
+    optional golem.rib.WorkerFunctionsInRib worker_functions_in_response = 11;
+    // type discriminator to keep backward compatibility
+    optional GatewayBindingType binding_type = 12;
+    optional StaticBinding static_binding = 13;
+    // middleware
+    optional Middleware middleware = 14;
```

Comment:
This is not a great data structure only at the edges but that is to keep the backward compatibility, the core logic uses data structures type safe.